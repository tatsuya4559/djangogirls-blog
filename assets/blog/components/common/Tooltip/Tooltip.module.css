.wrapper {
  display: inline-block;
  position: relative;
}

.tooltip {
  visibility: hidden;
  background-color: black;
  border-color: black;
  color: white;
  text-align: center;
  padding: 0.25rem 0.5rem;
  border-radius: 0.25rem;
  position: absolute;
  z-index: 1;
  display: block;
}

.wrapper:hover .tooltip {
  visibility: visible;
}

.top {
  bottom: calc(100% + 5px); /* 5px for the arrow */
}

.top::after {
  content: '';
  position: absolute;
  top: 100%;
  left: 50%;
  margin-left: -5px;
  border-width: 5px;
  border-style: solid;
  border-top-color: inherit;
  border-right-color: transparent;
  border-bottom-color: transparent;
  border-left-color: transparent;
}

.bottom {
  top: calc(100% + 5px); /* 5px for the arrow */
}

.bottom::after {
  content: '';
  position: absolute;
  bottom: 100%;
  left: 50%;
  margin-left: -5px;
  border-width: 5px;
  border-style: solid;
  border-top-color: transparent;
  border-right-color: transparent;
  border-bottom-color: inherit;
  border-left-color: transparent;
}

.right {
  top: 0;
  left: calc(100% + 5px); /* 5px for the arrow */
}

.right::after {
  content: '';
  position: absolute;
  top: 50%;
  right: 100%;
  margin-top: -5px;
  border-width: 5px;
  border-style: solid;
  border-top-color: transparent;
  border-right-color: inherit;
  border-bottom-color: transparent;
  border-left-color: transparent;
}

.left {
  top: 0;
  right: calc(100% + 5px); /* 5px for the arrow */
}

.left::after {
  content: '';
  position: absolute;
  top: 50%;
  left: 100%;
  margin-top: -5px;
  border-width: 5px;
  border-style: solid;
  border-top-color: transparent;
  border-right-color: transparent;
  border-bottom-color: transparent;
  border-left-color: inherit;
}
