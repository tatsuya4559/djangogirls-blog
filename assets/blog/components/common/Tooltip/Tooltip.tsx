import React from 'react';
import cx from 'classnames';
import classes from './Tooltip.module.css';

/**
 * 表示の微調整にclassNameを利用してください
 * placementが'top'または'bottom'の場合は左右の表示位置を調整してください
 */
type Props = {
  text: string;
  placement: 'top' | 'bottom' | 'right' | 'left';
  className?: string;
};

const Tooltip: React.FC<Props> = ({ text, placement, className, children }) => {
  return (
    <div className={classes.wrapper}>
      {children}
      <span className={cx(className, classes.tooltip, classes[placement])}>
        {text}
      </span>
    </div>
  );
};

export default Tooltip;
