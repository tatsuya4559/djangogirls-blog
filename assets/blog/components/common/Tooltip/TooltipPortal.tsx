import React, { useState } from 'react';
import { createPortal } from 'react-dom';

type TooltipProps = {
  text: string;
  onMouseEnter: () => void;
  onMouseLeave: () => void;
};

const Tooltip: React.FC<TooltipProps> = React.memo(
  ({ text, onMouseEnter, onMouseLeave }) => {
    return (
      <span onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave}>
        {text}
      </span>
    );
  },
);

type PortalProps = {
  text: string;
  placement: 'top' | 'bottom' | 'right' | 'left';
  delay?: number; // ms
  className?: string;
};

const TooltipPortal: React.FC<PortalProps> = React.memo(
  ({ text, placement, delay = 0, className, children }) => {
    const [visible, setVisible] = useState(false);
    const [timerId, setTimerId] = useState<number | undefined>();

    const onMouseEnter = () => {
      setTimerId(window.setTimeout(() => setVisible(true), delay));
    };
    const onMouseLeave = () => {
      window.clearTimeout(timerId);
      setVisible(false);
    };

    return createPortal(
      <Tooltip
        text={text}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
      >
        {children}
      </Tooltip>,
      document.body,
    );
  },
);

export default TooltipPortal;
