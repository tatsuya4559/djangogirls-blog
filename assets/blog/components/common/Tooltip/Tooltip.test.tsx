import React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Tooltip from './Tooltip';

afterEach(cleanup);

test('Snapshot', () => {
  const { asFragment } = render(
    <Tooltip text="tooltip" placement="top">
      Test this.
    </Tooltip>,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('textに指定した文字列が表示されていること', () => {
  render(
    <Tooltip text="this is test" placement="top">
      <button>hover me</button>
    </Tooltip>,
  );

  expect(screen.queryByText('this is test')).toBeTruthy();
});

test('childrenコンポーネントが表示されていること', () => {
  render(
    <Tooltip text="this is test" placement="top">
      <button>hover me</button>
    </Tooltip>,
  );

  expect(screen.queryByText('hover me')).toBeTruthy();
});
