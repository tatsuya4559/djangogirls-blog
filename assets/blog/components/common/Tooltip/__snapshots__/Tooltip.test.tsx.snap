// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    Test this.
    <span
      class="tooltip top"
    >
      tooltip
    </span>
  </div>
</DocumentFragment>
`;
